File: 13.TreeTraversals/BstTreeTraversals.c
#include <stdio.h>
#include <stdlib.h>

struct Node {
    int data;
    struct Node *left, *right;
};

int arr[100000];
int idx = 0;
int n = 0;

struct Node* buildTree() {
    if (idx >= n)
        return NULL;

    if (arr[idx] == -1) {
        idx++;
        return NULL;
    }

    struct Node* root = (struct Node*)malloc(sizeof(struct Node));
    root->data = arr[idx++];
    root->left = buildTree();
    root->right = buildTree();

    return root;
}

void preorder(struct Node* root) {
    if (root == NULL) return;
    printf("%d ", root->data);
    preorder(root->left);
    preorder(root->right);
}

void inorder(struct Node* root) {
    if (root == NULL) return;
    inorder(root->left);
    printf("%d ", root->data);
    inorder(root->right);
}

void postorder(struct Node* root) {
    if (root == NULL) return;
    postorder(root->left);
    postorder(root->right);
    printf("%d ", root->data);
}

int main() {
    while (scanf("%d", &arr[n]) != EOF) {
        n++;
    }

    struct Node* root = buildTree();

    preorder(root);
    printf("\n");
    inorder(root);
    printf("\n");
    postorder(root);

    return 0;
}

File: 13.TreeTraversals/BstTreeTraversals.py
# Write your code here!!

import sys
sys.setrecursionlimit(10**7)

class TreeNode:
    def __init__(self, value):
        self.id = value
        self.left = None
        self.right = None


data = list(map(int, sys.stdin.read().split()))
it = iter(data)


def build_tree():
    val = next(it)
    if val == -1:
        return None

    node = TreeNode(val)
    node.left = build_tree()
    node.right = build_tree()
    return node


def preorder(root):
    if root:
        print(root.id, end=" ")
        preorder(root.left)
        preorder(root.right)


def inorder(root):
    if root:
        inorder(root.left)
        print(root.id, end=" ")
        inorder(root.right)


def postorder(root):
    if root:
        postorder(root.left)
        postorder(root.right)
        print(root.id, end=" ")


root = build_tree()

preorder(root)
print()
inorder(root)
print()
postorder(root)
